scio-core/src/main/java/com/spotify/scio/transforms/BaseAsyncBatchLookupDoFn.java

```diff
@@ -0,0 +1,356 @@
+    } catch (Exception e) {
+      LOG.error("Failed to process element", e);
+      throw e;
+    }
+  }
+
+  @FinishBundle
+  public void finishBundle(FinishBundleContext context) {
+
+    // send remaining
+    try {
+      /** @todo handle exception properly * */
+      createRequest();
+      if (!futures.isEmpty()) {
+        // Block until all pending futures are complete
+        waitForFutures(futures.values());
+      }
+    } catch (InterruptedException e) {
+      Thread.currentThread().interrupt();
+      LOG.error("Failed to process futures", e);
+      throw new RuntimeException("Failed to process futures", e);
+    } catch (ExecutionException e) {
+      LOG.error("Failed to process futures", e);
+      throw new RuntimeException("Failed to process futures", e);
+    }
+    flush(r -> context.output(KV.of(r.input, r.output), r.timestamp, r.window));
+
+    // Make sure all requests are processed
+    Preconditions.checkState(
+        inputCount == outputCount,
+        "Expected requestCount == responseCount, but %s != %s",
+        inputCount,
+        outputCount);
+  }
+
+  private void createRequest() throws InterruptedException {
+    final ClientType client = getResourceClient();
+    final Cache<String, Output> cache = getResourceCache();
+    final UUID key = UUID.randomUUID();
+    final List<Input> elems = new ArrayList<>(batch);
+    final BatchRequest request = batchRequestFn.apply(elems);
+
+    // semaphore release is not performed on exception.
+    // let beam retry the bundle. startBundle will reset the semaphore to the
+    // maxPendingRequests permits.
+    semaphore.acquire();
+    final FutureType future = asyncLookup(client, request);
+    // handle cache in fire & forget way
+    handleCache(future, cache);
+    // make sure semaphore are released when waiting for futures in finishBundle
+    final FutureType unlockedFuture = handleSemaphore(future);
+
+    futures.put(key, handleOutput(unlockedFuture, elems, key));
+    batch.clear();
+  }
+
+  private FutureType handleOutput(FutureType future, List<Input> batchInput, UUID key) {
+    return addCallback(
+        future,
+        response -> {
+          batchResponseFn
+              .apply(response)
+              .forEach(
+                  pair -> {
+                    final String id = pair.getLeft();
+                    final Output output = pair.getRight();
+
+                    if (!inputs.containsKey(id)) {
+                      LOG.error(
+                          "The ID '{}' received in the gRPC batch response does not "
+                              + "match any IDs extracted via the idExtractorFn for the requested  "
+                              + "batch sent to the gRPC endpoint. Please ensure that the IDs returned "
+                              + "from the gRPC endpoints match the IDs extracted using the provided"
+                              + "idExtractorFn for the same input.",
+                          id);
```

Comment:
if this is a programming error, we should rather fail instead of logging

Reply:
I totally agree and that was actually my initial approach. But because we never act on this future after this point (no more callbacks to check if it's completed with a failure or success) i was unable to stop the execution here (or maybe i missed how to) So i opted here for the log. 
With that said if that log is emitted that means one `Result` will be missing from the bundle and the number or requests will not match the number of responses and the bundle will fail. 
If that's not good enough then maybe we could restructure the future handling fail immediately if the any of the output callbacks was not able to find the input.